File: Linked List/RemoveNthNodeFromEndOfList.java
class ListNode {
	int val;
	ListNode next;
	ListNode(int val) { this.val = val; }
}

// Use dummy to handle edges cases like deleting head. Initialize two pointers first and second at dummy. Move second n + 1 steps to have a headstart, while first starts at head. Then have move both pointers at the same pace until second reaches the end. First will land just before the node we want to delete: update first.next to first.next.next to remove the target node.
public class RemoveNthNodeFromEndOfList {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;

        ListNode first = dummy;
        ListNode second = dummy;
        for (int i = 0; i <= n; i++) {
            second = second.next;
        }
        while (second != null) {
            first = first.next;
            second = second.next;
        }
        first.next = first.next.next;
        return dummy.next;
    }
 
}

File: Linked List/Remove Nth Node From End of List/remove_nth_node_from_end_of_list.cpp
class Solution {
public:
    // Two nodes starting at a dummy, one with a headstart of n + 1. 
    // Move together until headstart node reaches end of list 
    // -> prev lands just before nth node.
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode();
        dummy->next = head;
        ListNode* prev = dummy;
        ListNode* curr = dummy;
        for (int i = 0; i < n + 1; ++i) {
            curr = curr->next;
        }

        while (curr) {
            prev = prev->next;
            curr = curr->next;
        }
        prev->next = prev->next->next;

        return dummy->next;
    } 
};
